### ISSUE TITLE
Remote interactive debugger doesn't work when remotely attaching to RHEL and Python 2.6 

### ISSUE THREAD

**Issue by KronosKoderS** [Microsoft/PTVS#2097]
I'm trying to remotely attached and debug a script.  The interactive debugger doesn't appear to work.  When issuing a command, I get a "is not connected please restart process" error.  I then proceed to select the "Restart" button on the interactive window and get an internal error that says I need to restart VS.

**Comment by KronosKoderS** [comment]
I'm able to set breakpoints and stop on them.  View the stack and also view the locals.  I can even "interact" with the remote process by making function calls using the "Watch" window.

**Comment by zooba** [comment]
Are you using Visual Studio 2015 or VS 2017? (There are some pretty significant differences between the two, so it'll help us narrow down the problem sooner.)

**Comment by KronosKoderS** [comment]
2015 Pro.  Ptvs 2.0... I believe.  I'm not at my work computer and won't be until monday to confirm.

**Comment by KronosKoderS** [comment]
I've confirmed the Versions:

VS 2015 Pro Update 1
PTVS 2.0

**Comment by KronosKoderS** [comment]
Update:

I can actually use the remote debugger when running the script from the command line.  However, we have another program (java gui based) that kicks off a script using a new thread.  When trying to connect to this thread it appears I cannot use the remote interactive debugger (i.e. send a `print "Hello World"`).  I am able to remotely debug like stepping through the code and viewing the call stack.  I can also pseudo interact with the remote process using the "watch" window.  However, I get a connection error when trying to use the interactive debugger.  It prompts me to reset, and doing so, causes another error.

**Comment by KronosKoderS** [comment]
@zooba just curious if you've been able to look into this at all.

**Comment by zooba** [comment]
Sorry, no I haven't. We've been very busy here with getting our VS 2017 release ready.

Are you sure it's PTVS 2.0 and not 2.2? The latest available for VS 2015 is [2.2.6](https://github.com/Microsoft/PTVS/releases/tag/v2.2.6) (which the VS installer will tell you needs Update 3, but it should be fine on Update 1).

If you're actually on 2.0, then you'll need to make sure you `pip install ptvsd==2.0` on the remote machine. Otherwise there'll be a version mismatch that may cause issues. But I'd strongly recommend updating as much as you can.

**Comment by KronosKoderS** [comment]
@zooba you are correct, I am using V2.2.  The remote computer that I'm debugging isn't connected to the internet (and can't easily be done).  I also don't have admin access to install new packages.  Instead, I've copied `ptvsd` from my local computer into a folder on the remote machine and added that folder to the `PYTHONPATH`.  I remember reading somewhere that this was ok to do if necessary.

**Comment by KronosKoderS** [comment]
@zooba is there any documentation on debugging ptvsd?  I was hoping to attach or enable more detailed debug statements or something on the remote machine to see if there were any blocked packets or something of that nature.

**Comment by KronosKoderS** [comment]
@zooba I've found the culprit!  I redirect sys.stdout and sys.stderr to a custom class (it outputs to our GUI interface).  If I disable this output, ptvsd interactive debugging appears to work.  

Do you have any suggestions on how to continue to redirect the output, but still support ptvsd?

**Comment by KronosKoderS** [comment]
I was finally able to figure out how to get this working.  I created a custom class that inherited from `_DebuggerOutput` and overrode the `__init__` and `write` methods:

```
class InfoWriter(_DebuggerOutput):
    def __init__:
        super(InfoWriter, self).__init__(sys.stdout, True)

    def write(self, buf):
        # custom code here
        super(InfoWriter, self).write(buf)
```

It's kind of a hackish way of getting it to work, but it works wonderfully now.

**Comment by zooba** [comment]
I guess we're relying on getting some output back, and when stdout and stderr are overridden without calling our class then it fails.

You might also be able to do something like this in your own class:

```
class InfoWriter:
    def __init__(self, write_to):
        self.write_to = write_to

    def write(self, buf):
        self.write_to.write(buf)
        # whatever you'd normally do

sys.stdout = InfoWriter(sys.stdout)
```

**Comment by KronosKoderS** [comment]
Only problem with that, is if we don't attach, then `write_to` would be `None`, and therefore would raise an exception.

**Comment by zooba** [comment]
Easily fixed with an if statement :)

I don't know that we have the ability to reliably retrieve output in this scenario. Potentially we could swap out `sys.stdout` and `stderr` around expressions, but I feel like that's likely to cause issues more frequently than it resolves them. We might be able to warn if they are changed though - I'll keep this issue around as a reminder to look into it.

**Comment by KronosKoderS** [comment]
Perhaps an alternative step would be to report this as a known issue and update documentation?
